lib/nerves_hub/console_channel.ex

```diff
@@ -0,0 +1,175 @@
+  end
+
+  def handle_in(
+        "io_response:" <> "put_chars",
+        %{"data" => data},
+        %{request: _, iex_pid: _} = state
+      ) do
+    {from, reply_as, {:put_chars, _, _}} = state.request
+    io_reply(from, reply_as, String.to_atom(data))
+    {:noreply, %{state | request: nil}}
+  end
+
+  def handle_in(
+        "io_response:" <> "get_line",
+        %{"data" => data},
+        %{request: _, iex_pid: _} = state
+      ) do
+    {from, reply_as, {:get_line, _, _}} = state.request
+    io_reply(from, reply_as, data)
+    {:noreply, %{state | request: nil}}
+  end
+
+  def handle_in(event, payload, state) do
+    {:stop, {:unhandled_event, event, payload}, state}
+  end
+
+  def handle_reply(
+        {:ok, :join, %{"response" => _response, "status" => "ok"}, _},
+        state
+      ) do
+    {:noreply, state}
+  end
+
+  def handle_reply(
+        {:error, :join, %{"response" => %{"reason" => reason}, "status" => "error"}},
+        state
+      ) do
+    {:stop, reason, state}
+  end
+
+  def handle_reply(_, state) do
+    {:noreply, state}
+  end
+
+  def handle_close(_payload, state) do
+    Process.send_after(self(), :rejoin, 5_000)
+    {:noreply, state}
+  end
+
+  # This is the group_leader message.
+  def handle_info({:io_request, from, reply_as, req}, state) do
+    state = io_request(from, reply_as, req, state)
+    {:noreply, state}
+  end
+
+  # Matches all the :io_request commands.
+  # This is not definitive. These are just the ones for
+  # Basic IEx interaction.
+
+  # :setopts not supported
+  defp io_request(from, reply_as, {:setopts, _opts}, state) do
+    reply = {:error, :enotsup}
+    io_reply(from, reply_as, reply)
+    state
+  end
+
+  # :getopts not supported
+  defp io_request(from, reply_as, :getopts, state) do
+    reply = {:ok, [binary: true, encoding: :unicode]}
+    io_reply(from, reply_as, reply)
+    state
+  end
+
+  # :get_geometry not supported
+  defp io_request(from, reply_as, {:get_geometry, :columns}, state) do
+    reply = {:error, :enotsup}
+    io_reply(from, reply_as, reply)
+    state
+  end
+
+  defp io_request(from, reply_as, {:get_geometry, :rows}, state) do
+    reply = {:error, :enotsup}
+    io_reply(from, reply_as, reply)
+    state
+  end
+
+  # All other requests push 
+  # a command over the socket. 
+  defp io_request(from, reply_as, req, state) do
+    # FIXME(Connor) 18-09-17
+    # This function needs to be spawned because
+    # it calls the `push` function, which 
+    # is a GenServer call. There may be a better solution for
+    # this.
+    request = {from, reply_as, req}
+    spawn_link(__MODULE__, :push_request, [from, reply_as, req])
+    Map.put(state, :request, request)
+  end
+
+  # Pushes an io_request onto the socket.
+  # will be published as event="io_request:" <> command with a map payload
+  # of the arguments. 
+  def push_request(_from, _reply_as, {:put_chars, :unicode, msg}) do
+    push("io_request:put_chars", %{data: msg, encoding: "unicode"})
```

Comment:
I think the `msg` here could be iodata, which probably won't work nicely with the json encoders. It probably would be a good idea to throw in a `IO.iodata_to_binary` in there.

Reply:
Good call. :+1: